fastcache_flush.go

```diff
@@ -0,0 +1,230 @@
+// Package fastcache implements fast in-memory cache.
+//
+// The package has been extracted from https://victoriametrics.com/
+package turbocache
+
+import (
+	"github.com/ShareChat/turbo-cache/internal/spinlock"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type flusher struct {
+	count               int
+	chunks              []flushChunk
+	chunkSynced         atomic.Value
+	index               atomic.Value
+	idx                 uint64
+	gen                 uint64
+	currentChunkId      uint64
+	currentFlushChunkId int32
+	totalChunkCount     uint64
+	needClean           bool
+	flushed             atomic.Bool
+	spinlock            spinlock.RWMutex
+}
+
+type flushChunk struct {
+	chunkId    uint64
+	chunk      [chunkSize]byte
+	h          []uint64
+	idx        []uint64
+	gen        []uint64
+	size       uint64
+	cleanChunk bool
+}
+
+func (b *bucket) onFlushTick(flushInterval int64) {
+	if b.flusher.count > 0 && time.Since(time.UnixMilli(b.latestTimestamp)).Milliseconds() >= flushInterval {
+		b.setBatch(b.flusher)
+		b.cleanFlusher(b.flusher)
+	}
+}
+
+func (b *bucket) onNewItem(i *queuedStruct, maxBatch int, flushInterval int64) {
+	defer releaseQueuedStruct(i)
+
+	f := b.flusher
+
+	index := b.flusher.index.Load().([]flushChunkIndexItem)
+	indexId := i.h % uint64(len(index))
+	duplicated := index[indexId].exists(i.h)
+	forceFlush := false
+	if !duplicated {
+		kvLength := uint64(4) + uint64(len(i.K)) + uint64(len(i.V))
+		idxNew, newChunk := f.incrementIndexes(kvLength)
+		if newChunk {
+			f.currentFlushChunkId++
+		}
+		if f.currentFlushChunkId <= int32(len(f.chunks)) {
+			flushChunk := &f.chunks[f.currentFlushChunkId]
+
+			lenBuf := makeKvLenBuf(i.K, i.V)
+			//f.spinlock.Lock()
+			copy(flushChunk.chunk[flushChunk.size:], lenBuf[:])
+			copy(flushChunk.chunk[flushChunk.size+4:], i.K)
+			copy(flushChunk.chunk[flushChunk.size+4+uint64(len(i.K)):], i.V)
+
+			//f.spinlock.Unlock()
+			flushChunk.h = append(flushChunk.h, i.h)
+			flushChunk.idx = append(flushChunk.idx, f.idx)
+			flushChunk.gen = append(flushChunk.gen, f.gen)
+			flushChunk.chunkId = f.currentChunkId
+			flushChunk.cleanChunk = newChunk
+
+			for j := range index[indexId].h {
+				if atomic.LoadUint64(&index[indexId].h[j]) == 0 {
+					atomic.StoreUint64(&index[indexId].currentIdx[j], flushChunk.size)
+					atomic.StoreInt32(&index[indexId].flushChunk[j], f.currentFlushChunkId)
+					atomic.StoreUint64(&index[indexId].h[j], i.h)
+					break
+				}
+			}
+
+			b.flusher.index.Store(index)
```

Comment:
Don't understand. So we get `index` as `b.flusher.index.Load().([]flushChunkIndexItem)`, and here we `Store` it back. But don't we modify the existing slice anyway? Which actually means that the content of `index` may be inconsistent during these non-atomic calls

```
if atomic.LoadUint64(&index[indexId].h[j]) == 0 {
					atomic.StoreUint64(&index[indexId].currentIdx[j], flushChunk.size)
					atomic.StoreInt32(&index[indexId].flushChunk[j], f.currentFlushChunkId)
					atomic.StoreUint64(&index[indexId].h[j], i.h)
					break
				}
```

?

Reply:
I will write it with details later. The key idea is:
1. we have chunk with ready to write data
2. we have index above it

during writing
1. add data to the chunk, just append bytes in array without recopying the array in a heap
2. append data in index (index item contains several `arrays`, not slices). the order makes sense here, firstly we add index to flush chunk, than hash of the key.

During **flushing** we use spin lock and clean index, than clean flush chunk itself

During reading from the index with firstly check hash in index, than we read bytes from so-called flush chunk. reading from flush chunks are executed under the spin lock.

Also I had a flag  about flushing and the read process doesnt acquire for the spin lock in this case.

So, basically we most of the time have append only writing, for reading bytes && cleaning the index with use spin lock

Reply:
it's relevant to your second comment as well. I think I need to write it down in description as well with some visualisation